Fix file lock issue in AotCodeCompiler

masnesral opened the pull request:
Summary:
It looks like there are several places in AotCodeCompiler that write files in a way that aren't safe for concurrency. There's a filelock to cope with that, but it seems like the lock path isn't quite robust enough to prevent races. We have an internal stress test failing when executing multiple concurrent versions of the test. It seems as though there's some variability in the content we write to the cpp file, which means we can get a different 'key' across different runs. The lock path includes that key in the lock path name, but the path for the "consts_path" is computed separately. Therefore, I see things like this:

- The computed 'key' is `cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z`
- The lock_path (based on the key) is: `/tmp/torchinductor_slarsen/locks/cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z.lock`
- The cpp path is (also includes the key) is: `/tmp/torchinductor_slarsen/cenzkqfnhu53mrhrdhzjtnblzyma2hgmeo7hai5yqsxzirdavurh/cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z.cpp`
- The consts_path (not based on the key) is: `/tmp/torchinductor_slarsen/cenzkqfnhu53mrhrdhzjtnblzyma2hgmeo7hai5yqsxzirdavurh/cifbshkqkbsurzldsyi2vl5bsnhvejmavys4kktpwrzmpo4ysuoy.bin`

So we have different test instances using different lock paths, but touching the same consts_path and therefore stomping on each others' consts_path. To fix, include the key in the consts_paths.

Test Plan: Ran internal stress test. Repro'd failure and verified this change fixes it.

Differential Revision: D60552021


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

desertfire reviewed (approved):
LGTM. Since specified_output_path can be specified by user with https://github.com/pytorch/pytorch/blob/6ff1e43a416c43cd82b210e22ac47384494c172e/torch/_inductor/config.py#L864, can you double check that also works?

masnesral commented:
> can you double check that also works?

Hmm, @desertfire maybe this isn't the way to solve it then. On your suggestion, when I set output_path="foobar", I get a lock file like:

`/tmp/torchinductor_slarsen/locks/foobar.lock`

It works, but is probably overly restrictive. So another way to fix this would be to go back to using "key" in the lock name, but make sure the `consts_path` also includes the key. WDYT?

masnesral committed:
Fix file lock issue in AotCodeCompiler (#132343)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/132343

It looks like there are several places in AotCodeCompiler that write files in a way that aren't safe for concurrency. There's a filelock to cope with that, but it seems like the lock path isn't quite robust enough to prevent races. We have an internal stress test failing when executing multiple concurrent versions of the test. It seems as though there's some variability in the content we write to the cpp file, which means we can get a different 'key' across different runs. The lock path includes that key in the lock path name, but the path for the "consts_path" is computed separately. Therefore, I see things like this:

- The computed 'key' is `cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z`
- The lock_path (based on the key) is: `/tmp/torchinductor_slarsen/locks/cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z.lock`
- The cpp path is (also includes the key) is: `/tmp/torchinductor_slarsen/cenzkqfnhu53mrhrdhzjtnblzyma2hgmeo7hai5yqsxzirdavurh/cp5tgbuxuegvg5g2j7oi6u74nkf3v7mx5w3qzl6qbedtmw5tq77z.cpp`
- The consts_path (not based on the key) is: `/tmp/torchinductor_slarsen/cenzkqfnhu53mrhrdhzjtnblzyma2hgmeo7hai5yqsxzirdavurh/cifbshkqkbsurzldsyi2vl5bsnhvejmavys4kktpwrzmpo4ysuoy.bin`

So we have different test instances using different lock paths, but touching the same consts_path and therefore stomping on each others' consts_path. To fix, include the key in the consts_paths.

Test Plan: Ran internal stress test. Repro'd failure and verified this change fixes it.

Differential Revision: D60552021

desertfire commented:
> but is probably overly restrictive.

Can you elaborate more? I think it is not too bad? If a user specify an output path, both .cpp and .bin files will be written to the same directory, and we need to make sure there is stomp in this case too.

masnesral commented:
@desertfire 

> Can you elaborate more?

Yeah, I just mean that if you have set output_path="foobar" then the method in question would lock on a path like: `/tmp/torchinductor_slarsen/locks/foobar.lock`, which means that ALL processes you run will compete for that same lock file, even though they would otherwise be writing different files (by virtue of the different 'key' string embedded in most of the file names). The problem that I've been encountering with the internal stress tests is that we may have two executions creating different key values (therefore distinct cpp files), but they create the SAME consts_path. The bug is that the operations on the consts_path are not guarded by this lock file.

masnesral commented:
The latest rev shows an alternative way to fix the problem, i.e., embed the key string in the consts_path.

masnesral commented:
@pytorchbot merge

masnesral commented:
@pytorchbot merge
